pastas/solver.py

```diff
@@ -657,3 +662,439 @@ def objfunction(
     ) -> ArrayLike:
         p = np.array([p.value for p in parameters.values()])
         return self.misfit(p=p, noise=noise, weights=weights, callback=callback)
+
+
+class EmceeSolve(BaseSolver):
+    """Solver based on MCMC approach in emcee :cite:p:`foreman-mackey_emcee_2013`.
+
+    Parameters
+    ----------
+    objective_function: func, optional
+        An objective function to be minimized. If not provided, the
+        GaussianLikelihood is used. See the pastas.objective_functions module for
+        more information.
+    nwalkers: int, optional
+        Number of walkers to use. Default is 20.
+    backend: emcee.backend, optional
+        One of the Backends from Emcee used to store MCMC results. See Emcee
+        for more information.
+    moves: emcee.moves, optional
+        The moves argument determines how the next step for a walker is chosen in
+        the MCMC approach. One of the Moves classes from Emcee has to be provided.
+        See Emcee documentation for more information.
+    parallel: bool, optional
+        Run the sampler in parallel or not.
+    progress_bar: bool, optional
+        Show the progress bar or not. Requires the `tqdm` package to be installed.
+    **kwargs, optional
+        All other keyword arguments are passed on to the BaseSolver class.
+
+    Notes
+    -----
+    The EmceeSolve solver uses the emcee package to perform a Markov Chain Monte Carlo
+    (MCMC) approach to find the optimal parameter values. The solver can be used as
+    follows:
+
+    >>> solver = ps.EmceeSolve(
+    ...     nwalkers=20,
+    ...     progress_bar=True,
+    ...     )
+    >>> ml.solve(solver=solver)
+
+    The arguments provided are mostly passed on to the `emcee.EnsembleSampler`
+    and determine how that instance is created. Arguments you want to pass on to
+    `run_mcmc` (and indirectly the `sample` method), can be passed on to
+    `Model.solve`, like:
+
+    >>> ml.solve(solver=ps.EmceeSolve(), thin_by=2)
+
+    Examples
+    --------
+
+    >>> ml.solve(solver=ps.EmceeSolve(), steps=5000)
+
+    To obtain the MCMC chains, use:
+
+    >>> ml.fit.sampler.get_chain(flat=True, discard=3000)
+
+    References
+    ----------
+    https://emcee.readthedocs.io/en/stable/
+
+    See Also
+    --------
+    emcee.EnsembleSampler
+    emcee.moves
+    emcee.backend
+    pastas.objective_functions
+
+    """
+
+    _name = "EmceeSolve"
+
+    def __init__(
+        self,
+        objective_function=None,
+        nwalkers: int = 20,
+        backend=None,
+        moves=None,
+        parallel: bool = False,
+        progress_bar: bool = True,
+        **kwargs,
+    ) -> None:
+        # Check if emcee is installed, if not, return error
+        try:
+            global emcee
+            import emcee as emcee  # Import emcee here, so it is no dependency
+        except ImportError:
```

Comment:
Is ImportError correct here, or will you run into ModuleNotFoundError if emcee is not installed?

Reply:
Nice cathch! I created a new Issue for this to solve in the next version, as it also applies to other parts of the code. #620 .